File: Data Structures & Algorithms/gas-station/submission-14.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int totalGas = 0;
        int totalCost = 0;
        for(int i = 0;i<gas.length;i++){
            totalGas += gas[i];
            totalCost += cost[i];
            gas[i] -= cost[i];
        }
        if(totalCost>totalGas){
            return -1;
        }
        int res = -1;
        int remainingGas = 0;
        for(int i=0;i<gas.length;i++){
            int g = gas[i];
            remainingGas +=g;
            if(remainingGas>=0 && res==-1){
                res = i;
            }
            else if(remainingGas<0){
                res = -1;
                remainingGas = 0;
            }
        }
        return res;

        // -2  -2  -2  3   3

    
        
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-14.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        smap = {}
        l,r = 0,0
        res = 0
        while r<len(s):
            if s[r] in smap:
                l = max(l,smap[s[r]] +1)
            smap[s[r]] = r
            res = max(res,r-l+1)
            r+=1
        return res

        
        